### ISSUE TITLE
Context.isClosed() and finalization

### ISSUE THREAD

**Issue by Daniel-Alievsky** [oracle/graal#4871]
I use org.graalvm.polyglot.Context class to work with JavaScript. It is auto-closeable, and, of course, it has "close()" method.

However, I didn't find "isClosed()" method or something like this. How can I know, whether a given context is already closed or no? For comparison, TruffleContext has such method.

Then, I didn't find any finalization mechanism inside Context class, like Cleaner (since Java 9) or finalize() method. However, in my software there are situations, when I cannot strictly guarantee closing all instances of Context - it depends on user's actions. (Our users can create Context instances in many "windows" to process JavaScript code and, maybe, share its variables between different windows.) 

Is it possible to add Cleaner mechanism into Context? Now I was enforced to create own class for this needs, GraalContextWrapper, which contains a reference to Context inside and performs necessary cleaning. It is the reason of my previous question - I don't want to close a context if it was already closed.

**Comment by oubidar-Abderrahim** [comment]
Thank you for your suggestion, we'll take a look into it

**Comment by Daniel-Alievsky** [comment]
Maybe my wrapper, that I already wrote, "GraalPerformer", can help. If solves not only the problem with cleaning, but also provide a concept of "auto-binding": some language, for which we will find bindings simultaneously with creating new context. It can help if we want to store a number of performers and access to variables quickly at any time (getBindings() method is not too quick). Of course, if we need bindings for other languages, different from chosen as "auto", it will not help, but in most typical situations we work with the same language.

```
public class GraalPerformer implements AutoCloseable {
    static final Logger LOG = Logger.getLogger(GraalPerformer.class.getName());

    private static final Cleaner CLEANER = Cleaner.create();

    private final ExpensiveCleanableState state;
    private final Cleaner.Cleanable cleanable;

    private GraalPerformer(Context context, String autoBindingLanguage) {
        this.state = new ExpensiveCleanableState(context, autoBindingLanguage);
        this.cleanable = CLEANER.register(this, state);
    }

    public static GraalPerformer newPerformer(Context context, String autoBindingLanguage) {
        return new GraalPerformer(context, autoBindingLanguage);
    }

    public Context context() {
        return state.context;
    }

    public String autoBindingLanguage() {
        return state.autoBindingLanguage;
    }

    public boolean hasAutoBindings() {
        return state.autoBindings != null;
    }

    public Value autoBindings() {
        return state.autoBindings;
    }

    public Value jsBindings() {
        return bindings(GraalSourceContainer.JAVASCRIPT_LANGUAGE);
    }

    public Value bindings(String language) {
        Objects.requireNonNull(language, "Null language for binding");
        return language.equals(state.autoBindingLanguage) ? autoBindings() : state.context.getBindings(language);
    }

    public Value perform(Source source) {
        return state.context.eval(source);
    }

    public Value perform(GraalSourceContainer sourceContainer) {
        return perform(sourceContainer.source());
    }

    public boolean isClosed() {
        return state.isReallyClosed();
    }

    @Override
    public void close() {
        state.normallyClosed = true;
        cleanable.clean();
    }

    @Override
    public String toString() {
        return state.toString();
    }

    void setContextId(Object contextId) {
        state.contextId = contextId;
    }

    private void checkClosed() {
        if (state.normallyClosed || state.isReallyClosed()) {
            // - will be true even WHILE executing cleanable.clean
            throw new IllegalStateException("Cannot use " + this);
        }
    }

    private static class ExpensiveCleanableState implements Runnable {
        private final String contextString;
        private final String autoBindingLanguage;
        private volatile Context context;
        private volatile Value autoBindings;
        private volatile boolean normallyClosed = false;
        private volatile Object contextId = null;
        // - optional, only for better toString

        ExpensiveCleanableState(Context context, String autoBindingLanguage) {
            this.context = Objects.requireNonNull(context, "Null context");
            this.contextString = context.toString();
            this.autoBindingLanguage = autoBindingLanguage;
            this.autoBindings = autoBindingLanguage == null ? null : context.getBindings(autoBindingLanguage);
        }

        @Override
        public void run() {
            if (isReallyClosed()) {
                return;
            }
            if (normallyClosed) {
                LOG.config(() -> "Normal closing " + toBriefString());
            } else {
                LOG.warning(() -> "CLEANING " + toBriefString());
            }
            context.close();
            context = null;
            autoBindings = null;
        }

        @Override
        public String toString() {
            return (!isReallyClosed() ? "" : normallyClosed ? "closed " : "abnormally CLEANED ") + toBriefString();
        }

        private String toBriefString() {
            return "Graal performer " +
                    "of " + contextString +
                    (contextId == null ? "" : " in context #" + contextId);
        }

        private boolean isReallyClosed() {
            return context == null;
        }
    }
}

```

**Comment by chumer** [comment]
> However, I didn't find "isClosed()" method or something like this. How can I know, whether a given context is already closed or no? 

The reason is that isClosed() and close() is fundamentally racy when used on multiple threads. The best approach here would be if we would just return a `boolean` for the close method whether it was actually closed. Then we can guarantee the return value only to return true if the close was actually performed for one thread once. Unfortunately, that would not be a compatible change, and we would need to add a new method for that.  

Another way of doing this would be to have a onClosing callback that you can provide in Context.Builder.

Would such a return value or callback for close also do the trick for you? Or would you need your wrapper anyway?

> (getBindings() method is not too quick)

Can you elaborate what you mean by that? getBindings() should be a fairly quick operation. If it's too slow, we can probably speed it up.

**Comment by Daniel-Alievsky** [comment]
> The reason is that isClosed() and close() is fundamentally racy when used on multiple threads. The best approach here would be if we would just return a boolean for the close method whether it was actually closed. Then we can guarantee the return value only to return true if the close was actually performed for one thread once. Unfortunately, that would not be a compatible change, and we would need to add a new method for that.

You are right. But, in this case, why did you provide isClosed() method in TruffleContext?
I think that the main purpose of isClosed method is logging (to warn if we forgot to close something). In this case multi-threading is not a serious problem. In any case, context may be closed in any time, even inside an operation. Even if you will provide more correct method 
boolean closeAndReturnPreviousState(),
what useful we can do if it returned "true", excepting logging?

>  Or would you need your wrapper anyway?

The main goal of my wrapper if cleaning resources while garbage collection. 

> Can you elaborate what you mean by that? getBindings() should be a fairly quick operation. If it's too slow, we can probably 
speed it up.

On my computer, after 10 iterations to warm up the JVM, creating new context, for example, Context.create(), requires ~2-3 ms. After this, calling context.getBindings("js") requires ~7-10 ms. Of course, only 1st call (further calls works quickly), but it is essential time, comparing with actual executing scripts, and I think it would be better to do this together with creating a context - at least, it will provide better timing.
